packages/yarnpkg-core/sources/Project.ts

```diff
@@ -703,9 +699,166 @@ export class Project {
+
+          // Convert our matches into descriptors
+          const matchDescriptors = matches.map(match => {
+            return templateVariantPattern(potentialVariants.pattern, match);
+          });
+
+          // Calculate the matches for the cache
+          const cacheMatches: Array<VariantParameters> = [];
+
+          for (const cacheParameters of cacheParameterMatrix) {
+            const mergedCacheParameters = {
+              ...thisPackageVariantParameters,
+              ...cacheParameters,
+            };
+            cacheMatches.push(...matchVariantParameters(possibilities, mergedCacheParameters, variantParameterComparators));
+          }
+
+          const cacheMatchDescriptors = cacheMatches.map(match => {
+            const matchParameters = pkgVersion ? {...match, version: pkgVersion} : match;
+            return templateVariantPattern(potentialVariants.pattern, matchParameters);
+          });
+
+          // In parallel, schedule resolution of all our cached descriptors
+          await Promise.all(
+            cacheMatchDescriptors.map(async matchDescriptor => {
+              try {
+                const alreadyResolved = descriptorResolutionPromises.has(matchDescriptor.descriptorHash);
+                const cacheEntry = await scheduleDescriptorResolution(matchDescriptor, variantParameters, workspace, pkgParent);
+
+                if (!alreadyResolved) {
+                  opts.report.reportInfo(MessageName.UNNAMED, `Adding variant cache entry:, ${
+                    structUtils.prettyLocator(this.configuration, pkg)
+                  } -> ${
+                    structUtils.prettyLocator(this.configuration, cacheEntry)
+                  }`);
+                }
+              } catch (resolveFailure) {
+                // Don't worry about it
+                opts.report.reportError(MessageName.UNNAMED, `Resolve failure for cache, ${
+                  resolveFailure
+                }`);
+              }
+            })
+          );
+
+          // For each potential match, try and resolve it, if it succeeds, stop searching
+          matchLoop: for (const matchDescriptor of matchDescriptors) {
+            try {
+              const pkgParentOrWorkspace = pkgParent ?? workspace;
+
+              const boundMatchDescriptor = resolver.bindDescriptor(matchDescriptor, pkgParentOrWorkspace, resolveOptions);
+
+              // Find the old dependency in our parent
+              parentDependencyLoop: for (const [parentDependencyPkgIdentHash, parentDependencyPkgDescriptor] of pkgParentOrWorkspace.dependencies) {
+                if (parentDependencyPkgDescriptor.name === pkg.name && parentDependencyPkgDescriptor.scope === pkg.scope) {
+                  // Use the VariantRemapResolver
+                  const remapDescriptor = structUtils.makeDescriptor(
+                    parentDependencyPkgDescriptor,
+                    `variant:${structUtils.stringifyDescriptor(boundMatchDescriptor)}`
```

Comment:
If I build this with the structUtils helper, it html encodes characters that mean the resolver can't grab the matched descriptor.

Reply:
`npm:variant-fallback@1.0.0` was being converted to `npm%3Avariant-fallback@1.0.0`